### ISSUE TITLE
Document how Campaigns expire

### ISSUE THREAD

**Issue by aaronschachter** [DoSomething/product#43]
How does the "5 days left" get set?

**Comment by marahml** [comment]
currently being discussed on slack. @jonuy can you update this once a decision is made? would be great if we could automate in some way / this wouldn't need to be done manually

**Comment by mikefantini** [comment]
@aaronschachter @marahml @jonuy shouldn't this be set the same way it is for campaigns leaving the hot phase on the site? 

in this case we're using high season end date on the campaign node. related issue: https://github.com/DoSomething/dosomething/issues/4444

**Comment by aaronschachter** [comment]
Hi @mikefantini ! The process of scheduling campaigns (if they are actually scheduled and not just set with the "In mobile app" flag) for the mobile app something we need to document, for sure.  Asked @sbsmith86 about this in Slack:

[1:45 PM] Shae Smith: @aaronschachter: @uy Campaigns will run longer than a month on the web, but they will be in the “hot phase” for only a month
[1:45 PM] Shae Smith: when they are in the hot phase is when we show the hot module
[1:45 PM] Shae Smith: and promote them more and stuff
[1:46 PM] Shae Smith: so Hawkeye aims to manage the schedule of when things are in and out of their hot phase. It sounds like the mobile app will only be pulling in the campaigns that are in their hot phase?

@marahml @jonuy  -- Can you help answer @mikefantini 's question?

**Comment by mikefantini** [comment]
haha hey!

I guess I just assumed that we would have a rule like "show in mobile app if campaign a. "on mobile app" is selected and b. if campaign is in high season.

so for instance if Backpack Attack was "on mobile app" and high season was set for August 1st to August 31st then it would appear in the app for the month of August and then no longer be in the app starting September 1st.

part of the goal of hawkeye is to account for the needs of the mobile app when it comes to scheduling to make sure that we have enough campaigns for the mobile app in any given month.

and this is part of the idea of making something hot, that it is when DS is putting it's organizational energy around the campaign - including featuring it on the mobile app.

![image](https://cloud.githubusercontent.com/assets/6330971/8704282/08f76fa2-2af0-11e5-9950-495d6c9affcd.png)

![image](https://cloud.githubusercontent.com/assets/6330971/8704300/1db5cd3a-2af0-11e5-89f5-d04a5ae95eec.png)

**Comment by jonuy** [comment]
ooh if high season / hot phase is only for a month, then that certainly can be a potential solution. One area that we'll need to check though is with the mobile app campaign calendar Marah and Felicia have worked through. As long as the months campaigns are scheduled to go on the app matches up with their hot phase, then we should be good.

**Comment by aaronschachter** [comment]
@uy - Wouldn't we need to support filtering the `GET campaigns` query by the High Season start/end date when getting campaigns?

**Comment by marahml** [comment]
this works as long as we're OK with putting campaigns in high season on the site, that are already set to be in the mobile app. 

see current content schedule: https://docs.google.com/spreadsheets/d/172ERW-byKua2XL0btQsL54MmvEuKZtxPGqi5giHnP6I/edit#gid=618503661

**Comment by jonuy** [comment]
I don't think we'd need to add filtering by high season. The `on_mobile_app` flag can still act as the main thing to filter against. The high season dates in the API response then really just act as fields for the app to reference for timing things.

**Comment by aaronschachter** [comment]
Couldn't we filter by high season to make it easier to find which campaigns should be active in the app though?

**Comment by jonuy** [comment]
I might be missing something, but filtering by high season alone I don't think would be enough since there will potentially be other campaigns in high season that aren't meant for the mobile app (campaigns focused on SMS games for example).

**Comment by aaronschachter** [comment]
Sorry - I meant additionally filtering by high season, as well as `mobile_app=true`. Something like GET `campaigns?mobile_app=true&startdate>timestart&enddate<endstart`. Potentially even just a flag instead, like `high_season=true`, to avoid each client calculating what high season should be.

I think this would allow us to schedule the campaigns in advance (instead of manually updating and saving the Mobile App field in order to publish/publish from the app).

**Comment by marahml** [comment]
I'm still not clear on this aspect: 

There are totally different campaigns that we will be promoting in the app, on a totally different schedule than the DS US sites high season campaigns. 

Is there a way to have "high season" campaigns on the site that users dont see as high season if you guys are on a different schedule? @mikefantini

**Comment by mikefantini** [comment]
ok sorry I didn't mean to complicate things. last comment I'll make and then let you guys make the call.

@marahml I didn't fully realize that 1. you guys had campaigns set for the first few months and 2. that there were 16 campaigns at any given time in the app (I may have known this at one time but must have forgotten). The reason I'd be hesitant about fully linking these 16 campaigns as being hot with the site is that we have different criteria outside of the interest groups you guys are trying to balance on the app side (i.e. sports, fashion, etc.) and have to make sure on the web site that the 15-20 hot campaigns each month have diversity between cause, type, time.

So given that, it doesn't seem like using the definition of on mobile app and in high season makes sense right now.

It would be nice to come up with another solution so that an admin doesn't have to unselect 16 campaigns as being `on_mobile_app` and select another 16 to have that distinction at midnight the first day of every month.

Perhaps this is something hawkeye can help with in the future, but I'm not sure when that will have the ability to manipulate data on the drupal side.

If I can help with this further let me know, but otherwise I'll stay out of it as to not cause further confusion.

**Comment by aaronschachter** [comment]
@marahml @jonuy  - If I recall correctly, in Android it sounds like TouchLab sets an expiration date when a campaign is first saved to the phone from the API. Doesn't this mean that the expiration date will change from user to user ?  Or does it actually pick a calendar date to expire?

**Comment by jonuy** [comment]
I haven't looked at the code. But if this is what they're doing, it'd make sense to pick a calendar date to expire at since we're setting the expectation that campaigns will no longer be available on the app at the end of a month.

**Comment by jonuy** [comment]
Based on what we just mentioned during the stand up, I'll talk to Andrea or Matt about their thoughts on adding in a new field to campaigns where we can set end (and start?) date for campaigns on the mobile app.

**Comment by aaronschachter** [comment]
I do think this would be MUCH better than having a hardcoded calendar date. This also gives us flexibility to spread the campaign dates out too, so potentially not every campaign would expire at the same time.

We just need to add a single Drupal Date field, perhaps something like `field_mobile_app_date` that collects a both start and end date, like `field_high_season` and `field_low_season` do. Potentially we could/should just remove the Boolean `field_mobile_app` flag entirely to keep the codebase clean.

**Comment by marahml** [comment]
hey @jonuy - just wanted to check in if we're moving forward with this route

**Comment by aaronschachter** [comment]
I know we've added a field to Phoenix that has start/end dates, but not sure of its deployment status yet. @jonuy @weerd Do we also need a separate GH issue to expose the field to the Phoenix API?

**Comment by weerd** [comment]
@chloealee added a `mobile_app_dates` field to the DB and exposed on the CMS interface for campaigns, and it's also been added to the API for full campaign view :)

With no mobile app dates defined:

```
mobile_app: {
  dates: null
},
```

With mobile app dates defined:

```
mobile_app: {
  dates:
    start: "2014-03-18T00:00:00+0000",
    end: "2014-06-20T00:00:00+0000"
  }
},
```

**Comment by aaronschachter** [comment]
@jonuy Don't we also need to support query string parameters to figure out which campaigns to display? Or were you thinking we still just query for `mobile_app=true` ?

**Comment by aaronschachter** [comment]
Also, once this is pushed to prod, @jonuy can you update this doc: https://github.com/DoSomething/api/wiki/Mobile-App-Usage#what-does-the-campaign-response-map-to-in-the-app

**Comment by weerd** [comment]
Yeah, I think you would just `?mobile_app=true` and then check the `mobile_app {...}` in the returned response of campaigns. Unless there's some weird overlap that we haven't considered that would give weird results from `?mobile_app=true`...

**Comment by jonuy** [comment]
So we could keep the `mobile_app=true` flag, which would require some manual work to check/uncheck the box on campaigns each month we'd flip the campaigns.

What would probably be more ideal is modifying the behavior of that `mobile_app` flag so that instead of just checking against a checkbox, the logic were that `mobile_app=true` would only give campaigns that are live within the date parameters set on the `mobile_app` start/end date fields.

@weerd thoughts on the above?

**Comment by aaronschachter** [comment]
Right... but we don't know how far into the future we want to query. Down the road we might want to query for "upcoming campaigns" that aren't within the timeframe... but that also brings up questions around published, and also the request to have campaigns that are available only in the app and not on the web. @jonuy @weerd do we have an issue anywhere for how to handle making a campaign app-only ? Is that slated for v1?

**Comment by jonuy** [comment]
Added a Trello card to our backlog so that the rest of the team is aware that this is work we'll likely need to do: https://trello.com/c/finngIMs
